File: src/sFOM.m
function [err, final_it, appr] = sFOM(A, b, f, num_it, trunc_len, mgs, ex, tol, verbose)

    %{
    Function for performing the sketched FOM iterations.
    
    Outputs the errors (optional; if ex is false, err will be false), 
    the final iteration count, and the final approximation to f(A)b.
    %}

    % Initializations
    N = size(A,1);
    err = zeros(num_it,1);
    final_it = num_it;
    
    % Set up the sketching
    hS = setup_sketching_handle(N,2*num_it); % s = 2*m_max

    % Allocate for truncated Krylov basis
    V = zeros(N, trunc_len); % truncated orthonormal basis (latest trunc_len vectors)
    
    % Normalize v
    [v,~] = qr(b,0);
    
    % Add first vector to Krylov basis
    V = [ V(:,2:end) , v ];
    
    % Sketch v and add to sketched basis 
    SV = hS(v); s = size(SV,1);
    SV = [ SV, zeros(s,num_it) ];
    
    % Allocate for sketched AV
    SAV = zeros(s,num_it+1); 
    
    % Allocate for full Krylov basis 
    Vfull = zeros(N,num_it+1);
    Vfull(:,1) = v;

    % Do sFOM iters
    for m = 1:num_it
    
        % Extract latest Arnoldi vector
        w = V(:,end);
    
        Aw = A*w; 

        % Sketch Aw, and save 
        SAV(:,m) = hS(Aw); 

        % Update w
        w = Aw;
    
        % Orthogonalize
        if mgs % modified gram schmidt
            for i = 1:trunc_len
                w = w - V(:,i)*(V(:,i)'*w);
            end
        else % classical gram schmidt
            w = w - V*(V'*w);
        end
    
        % Normalize w
        [v,~] = qr(w,0);
    
        % Add to truncated basis, and discard vectors below truncation length
        V = [ V(:,2:end) , v ];
        
        % Update sketched V
        SV(:,m+1) = hS(v); 
        
        % Save v in full Krylov basis
        Vfull(:,m+1) = v;

        % Whiten basis
        [SVw, SAVw, Rw] = whiten_basis(SV(:,1:m), SAV(:,1:m));

        % Compute sFOM approx
        
        % Save previous ym for stopping crit
        % Hack
        if m>=2
            ym_prev = ym;
        end

        SVm = SVw;
        M = SVm'*SVm;  
        coeffs = M\(f( (SVm'*SAVw)/M )*(SVm'*hS(b)));
        ym = (Rw\coeffs);
        appr = Vfull(:,1:m)*ym;

        % Get error
        if ex ~= false
            err(m) = norm(appr - ex);
        end

        % Evaluate stopping criterion
        % Hack
        if m>=2
            stop_crit = norm(Vfull(:,m)) / norm(SV(:,m));
            stop_crit = stop_crit * norm( SV(:,1:m) * (ym - [ym_prev;0]) );
    
            if stop_crit < tol
                final_it = m;
                err = err(1:m);
                if verbose
                    disp(strcat("Converged to within tolerance in ",...
                        num2str(final_it), " iterations. Final estimate on iter-diff: ", num2str(stop_crit)))
                end
                break
            end

        end
    
    end
    
    if final_it == num_it
        if verbose
            disp(strcat("Warning! Did not converge to within tolerance. Final estimate on iter-diff: ", num2str(stop_crit)))
        end
    end
    
    if ex == false
        err = false;
    end
end

File: src/sFOM.jl
"""
    sFOM(A, b, f, num_it, trunc_len, mgs, iter_diff_tol, sketch)

Compute the sFOM approximation `fₘ≈f(A)b`.

# Arguments
- `A::Matrix`
- `b::Vector`
- `f::Function`
- `num_it::Int` Max number of iterations
- `trunc_len::Int` Arnoldi truncation length
- `mgs::Bool` Use MGS orthogonalization, otherwise use GS
- `iter_diff_tol::Float` Iteration-difference tolerance
- `sketch::Function` Sketching function in sFOM
"""
function sFOM(A, b, f, num_it, trunc_len, mgs, iter_diff_tol, sketch)

    # Convergence flag
    conv_flag = -1;

    # Initializations
    N = size(A, 1);
    iter_diff = zeros(num_it);

    # Allocate for truncated Krylov basis
    V = zeros(ComplexF64, N, trunc_len);

    # Add first Krylov vector
    v = b/norm(b);
    V = hcat(V[:,2:end], v);

    # Sketch v and add to sketched basis
    SV = sketch(v);
    sketch_size = size(SV,1);
    SV = hcat(SV, zeros(ComplexF64, sketch_size, num_it));

    # Allocate for sketched AV
    SAV = zeros(ComplexF64, sketch_size, num_it+1);

    # Allocate for full Krylov basis
    Vfull = zeros(ComplexF64, N, num_it+1);
    Vfull[:,1] = v;

    # init approx, qm
    approx = zeros(size(b));
    qm = 0;

    # Do sFOM iters
    for m = 1:num_it

        # Compute mat-vec product with latest Arnolid vec
        v = V[:,end];
        Av = A*v;

        # Sketch Av and add to sketched basis 
        SAV[:,m] = sketch(Av);

        # Update v
        v = Av;

        # Orthogonalize
        if mgs
            for i=1:trunc_len
                v = v - V[:,i]*(V[:,i]'*v);
            end
        else
            v = v - V*(V'*v);
        end

        # Normalize 
        v = v/norm(v);

        # Add to truncated basis, and discard vectors below trunc_len
        V = hcat(V[:,2:end], v);

        # Update sketched V
        SV[:,m+1] = sketch(v);

        # Save v in full Krylov basis
        Vfull[:,m+1] = v;

        # Whiten basis
        SVw, SAVw, Rw = whitenBasis(view(SV,:,1:m), view(SAV,:,1:m));

        # Save previous approx
        if m >= 2
            qm_prev = qm;
        end

        # Compute sFOM qₘ
        SVm = SVw;
        M = SVm'*SVm;
        coeffs = M\( f((SVm'*SAVw)/M)*(SVm'*sketch(b)) );
        qm = (Rw\coeffs);

        # Evaluate stopping criterion 
        if m >= 2

            stop_crit = 1 / norm(SV[:,m]);
            stop_crit *= norm(SV[:,1:m]*(qm - vcat(qm_prev, 0)));
            iter_diff[m] = stop_crit;

            if stop_crit < iter_diff_tol

                # Compute full approx
                approx = view(Vfull,:,1:m)*qm;
                iter_diff = iter_diff[1:m];
                conv_flag = 1;

                return approx, conv_flag, iter_diff;
            end

        end

    end

    approx = view(Vfull,:,1:num_it)*qm;
    
    return approx, conv_flag, iter_diff;

end
